volmigrate/volmigrate.go

```diff
@@ -0,0 +1,170 @@
+//       Rather than creating a one-off package for them, they are intentionally directly copied.
+//       If another app like volcli/volmigrate is added in the future, we should extract these
+//       to a separate package.
+
+func errExit(ctx *cli.Context, err error, help bool) {
+	fmt.Fprintf(os.Stderr, "\nError: %v\n\n", err)
+	if help {
+		cli.ShowAppHelp(ctx)
+	}
+	os.Exit(1)
+}
+
+func execCliAndExit(ctx *cli.Context, f func(ctx *cli.Context) (bool, error)) {
+	if showHelp, err := f(ctx); err != nil {
+		errExit(ctx, err, showHelp)
+	}
+}
+
+func errorInvalidArgCount(rcvd, exptd int, args []string) error {
+	return errored.Errorf("Invalid number of arguments: expected %d but received %d %v", exptd, rcvd, args)
+}
+
+// -------------------------------------------------------------------------------------------------
+
+func latestMigrationVersion() int64 {
+	return availableMigrations[len(availableMigrations)-1].Version
+}
+
+func promptBeforeRunning(ctx *cli.Context, msg string) {
+	if ctx.GlobalBool("silent") {
+		return
+	}
+
+	fmt.Println(msg)
+	fmt.Println("Are you sure you want to continue? y/n")
+
+	reader := bufio.NewReader(os.Stdin)
+	text, err := reader.ReadString('\n')
+	if err != nil {
+		log.Fatalf("Failed to read input from stdin: %s", err)
+	}
+
+	text = strings.ToLower(text)
+	text = text[:len(text)-1] // strip newline
+
+	if text != "yes" && text != "y" {
+		fmt.Println("Aborting.")
+		os.Exit(1)
+	}
+}
+
+func List(ctx *cli.Context) {
+	execCliAndExit(ctx, list)
+}
+
+func list(ctx *cli.Context) (bool, error) {
+	if len(ctx.Args()) != 0 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 0, ctx.Args())
+	}
+
+	for _, m := range availableMigrations {
+		fmt.Printf("%d - %s\n", m.Version, m.Description)
+	}
+
+	return false, nil
+}
+
+func Run(ctx *cli.Context) {
+	execCliAndExit(ctx, run)
+}
+
+func run(ctx *cli.Context) (bool, error) {
+	e := etcd2.New(ctx.GlobalString("prefix"), ctx.GlobalStringSlice("etcd"))
+
+	if len(ctx.Args()) > 0 {
+		i, err := strconv.Atoi(ctx.Args()[0])
+		if err != nil {
+			return true, errored.Errorf("Invalid migration version: %v", ctx.Args()[0])
+		}
+
+		promptBeforeRunning(ctx, fmt.Sprintf("You have requested to run all pending migrations up to and including #%d.", i))
+		return runMigrationsUpTo(e, int64(i))
+	}
+
+	promptBeforeRunning(ctx, "You have requested to run all pending migrations.")
+	return runAllMigrations(e)
+}
+
+func runAllMigrations(b backend.Backend) (bool, error) {
+	return runMigrationsUpTo(b, latestMigrationVersion())
+}
+
+func findMigrationByVersion(cfg *config.Client, version int64) (*Migration, bool) {
+	for _, m := range availableMigrations {
+		if m.Version == version {
+			return m, true
+		}
+	}
+
+	return nil, false
+}
+
+func runMigrationsUpTo(b backend.Backend, targetVersion int64) (bool, error) {
+	migrationsRun := 0
+
+	currentVersion := b.CurrentSchemaVersion()
+
+	for _, m := range availableMigrations {
```

Comment:
Based on the other comment I just left, once I enforce that the next migration being registered has the expected sequential ID, this is guaranteed to be in order.  Right now, it's manually sorted in `volmigrate.init()`.  That and the `Migrations` type can be deleted once I make this change.